### PR TITLE
[Feature] add GroupNorm and LayerNorm\BatchNorm1D\BatchNorm2D\BatchNorm3D

### PR DESCRIPTION
**Pull request by justld**
[Feature] add GroupNorm and LayerNorm\BatchNorm1D\BatchNorm2D\BatchNorm3D

### CODE REVIEW

**Comment by justld** [comment]
test torch code:
```python
import torch
import numpy as np

np.random.seed(10001)
x = np.random.randn(4, 50, 244, 244).astype('float32')
print(x.min(), x.max())
x = torch.from_numpy(x)

l_o = torch.nn.GroupNorm(num_groups=5, num_channels=50, eps=1e-05, affine=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

l_o = torch.nn.LayerNorm(normalized_shape=x.shape[1:], eps=1e-05, elementwise_affine=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

l_o = torch.nn.GroupNorm(num_groups=5, num_channels=50, eps=1e-05, affine=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

l_o = torch.nn.LayerNorm(normalized_shape=x.shape[1:], eps=1e-05, elementwise_affine=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

```

paddle code:
```python
import paddle
import numpy as np
np.random.seed(10001)
x = np.random.randn(4, 50, 244, 244).astype('float32')
print(x.min(), x.max())
x = paddle.to_tensor(data=x)
l_o = paddle.nn.GroupNorm(num_groups=5, num_channels=50, epsilon=1e-05,
    weight_attr=None, bias_attr=None, data_format='NCHW', name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
l_o = paddle.nn.LayerNorm(normalized_shape=x.shape[1:], epsilon=1e-05,
    weight_attr=None, bias_attr=None, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
>>>l_o = torch.nn.GroupNorm(num_groups=5, num_channels=50, eps=1e-05, affine=
    False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
>>>l_o = torch.nn.LayerNorm(normalized_shape=x.shape[1:], eps=1e-05,
    elementwise_affine=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

```

**Comment by justld** [comment]
image

File: paconvert/api_matcher.py [hunk 1]
```diff
@@ -508,3 +508,109 @@ def generate_code(self, kwargs):
         )
         code = API_TEMPLACE.format(kwargs['weight'], kwargs['ignore_index'], reduction)
         return code
+
+
+class LayerNormMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
```
Comment (zhouwei25) [inline paconvert/api_matcher.py:514]: 这个elementwise_affine有转写方式吗

File: paconvert/api_matcher.py [hunk 2]
```diff
@@ -508,3 +508,109 @@ def generate_code(self, kwargs):
...
+                                name=None)
+            '''
+        )
+        code = API_TEMPLACE.format(kwargs['normalized_shape'], kwargs['eps'])
+        return code
+
+
+class GroupNormMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
+        if 'affine' in kwargs:
```
Comment (zhouwei25) [inline paconvert/api_matcher.py:534]: affine可以通过paddle.ParamAttr(learning_rate=0.0)来转写吧，映射文档里有描述

File: paconvert/api_matcher.py [hunk 3]
```diff
@@ -508,3 +508,109 @@ def generate_code(self, kwargs):
...
+                                data_format='NCHW', 
+                                name=None)
+            '''
+        )
+        code = API_TEMPLACE.format(kwargs['num_groups'], kwargs['num_channels'], kwargs['eps'])
+        return code
+
+
+class BatchNorm1DMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
```
Comment (zhouwei25) [inline paconvert/api_matcher.py:554]: 问题同上

File: paconvert/api_matcher.py [hunk 4]
```diff
@@ -508,3 +508,109 @@ def generate_code(self, kwargs):
...
+                                  use_global_stats={},
+                                  name=None)
+            '''
+        )
+        code = API_TEMPLACE.format(kwargs['num_features'], kwargs['momentum'], kwargs['eps'], kwargs['track_running_stats'])
+        return code
+
+
+class BatchNorm2DMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
```
Comment (zhouwei25) [inline paconvert/api_matcher.py:576]: 问题同上

File: paconvert/api_matcher.py [hunk 5]
```diff
@@ -508,3 +508,109 @@ def generate_code(self, kwargs):
...
+                                  use_global_stats={},
+                                  name=None)
+            '''
+        )
+        code = API_TEMPLACE.format(kwargs['num_features'], kwargs['momentum'], kwargs['eps'], kwargs['track_running_stats'])
+        return code
+
+
+class BatchNorm3DMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
```
Comment (zhouwei25) [inline paconvert/api_matcher.py:598]: 问题同上

**Review by zhouwei25 — COMMENTED** [review]
原则上尽量不要返回None (表示不支持转换)，影响论文复现。
除非没有办法替代或组合实现的，确定是功能缺失，才可暂时返回None（后面可能会收集这些地方二次开发）

File: paconvert/api_matcher.py (hunk 2 above)
Comment (justld) [reply paconvert/api_matcher.py]: Done

File: paconvert/api_matcher.py (hunk 3 above)
Comment (justld) [reply paconvert/api_matcher.py:572]: Done

File: paconvert/api_matcher.py (hunk 4 above)
Comment (justld) [reply paconvert/api_matcher.py:605]: Done

File: paconvert/api_matcher.py (hunk 5 above)
Comment (justld) [reply paconvert/api_matcher.py:638]: Done

**Comment by justld** [comment]
torch code:
```python
import torch
import numpy as np

np.random.seed(10001)
x = np.random.randn(4, 50, 244, 244).astype('float32')
print(x.min(), x.max())
x = torch.from_numpy(np.random.randn(4, 50, 244, 244).astype('float32'))

l_o = torch.nn.GroupNorm(num_groups=5, num_channels=50, eps=1e-05, affine=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

l_o = torch.nn.LayerNorm(normalized_shape=x.shape[1:], eps=1e-05, elementwise_affine=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244).astype('float32'))
l_o=torch.nn.BatchNorm1d(num_features=50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244, 244).astype('float32'))
l_o=torch.nn.BatchNorm2d(num_features=50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 24, 24, 24).astype('float32'))
l_o=torch.nn.BatchNorm3d(num_features=50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244).astype('float32'))
l_o=torch.nn.BatchNorm1d(num_features=50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244, 244).astype('float32'))
l_o=torch.nn.BatchNorm2d(num_features=50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 24, 24, 24).astype('float32'))
l_o=torch.nn.BatchNorm3d(num_features=50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244).astype('float32'))
l_o=torch.nn.BatchNorm1d(num_features=50, eps=1e-05, momentum=0.1, affine=False, track_running_stats=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244, 244).astype('float32'))
l_o=torch.nn.BatchNorm2d(num_features=50, eps=1e-05, momentum=0.1, affine=False, track_running_stats=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 24, 24, 24).astype('float32'))
l_o=torch.nn.BatchNorm3d(num_features=50, eps=1e-05, momentum=0.1, affine=False, track_running_stats=True, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = torch.nn.GroupNorm(num_groups=5, num_channels=50, eps=1e-05, affine=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

x = torch.from_numpy(np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = torch.nn.LayerNorm(normalized_shape=x.shape[1:], eps=1e-05, elementwise_affine=False, device=None, dtype=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

```


paddle code:
```python
import paddle
import numpy as np
np.random.seed(10001)
x = np.random.randn(4, 50, 244, 244).astype('float32')
print(x.min(), x.max())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = paddle.nn.GroupNorm(num_groups=5, num_channels=50, epsilon=1e-05,
    weight_attr=None, bias_attr=None, data_format='NCHW', name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
l_o = paddle.nn.LayerNorm(normalized_shape=x.shape[1:], epsilon=1e-05,
    weight_attr=None, bias_attr=None, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244).astype('float32'))
l_o = paddle.nn.BatchNorm1D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=None, bias_attr=None, data_format='NCL',
    use_global_stats=True, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = paddle.nn.BatchNorm2D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=None, bias_attr=None, data_format='NCHW',
    use_global_stats=True, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 24, 24, 24).astype('float32'))
l_o = paddle.nn.BatchNorm3D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=None, bias_attr=None, data_format='NCDHW',
    use_global_stats=True, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244).astype('float32'))
l_o = paddle.nn.BatchNorm1D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=None, bias_attr=None, data_format='NCL',
    use_global_stats=False, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = paddle.nn.BatchNorm2D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=None, bias_attr=None, data_format='NCHW',
    use_global_stats=False, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 24, 24, 24).astype('float32'))
l_o = paddle.nn.BatchNorm3D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=None, bias_attr=None, data_format='NCDHW',
    use_global_stats=False, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244).astype('float32'))
l_o = paddle.nn.BatchNorm1D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=paddle.ParamAttr(learning_rate=0.0), bias_attr=
    paddle.ParamAttr(learning_rate=0.0), data_format='NCL',
    use_global_stats=True, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = paddle.nn.BatchNorm2D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=paddle.ParamAttr(learning_rate=0.0), bias_attr=
    paddle.ParamAttr(learning_rate=0.0), data_format='NCHW',
    use_global_stats=True, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 24, 24, 24).astype('float32'))
l_o = paddle.nn.BatchNorm3D(num_features=50, momentum=1 - 0.1, epsilon=
    1e-05, weight_attr=paddle.ParamAttr(learning_rate=0.0), bias_attr=
    paddle.ParamAttr(learning_rate=0.0), data_format='NCDHW',
    use_global_stats=True, name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = paddle.nn.GroupNorm(num_groups=5, num_channels=50, epsilon=1e-05,
    weight_attr=paddle.ParamAttr(learning_rate=0.0), bias_attr=paddle.
    ParamAttr(learning_rate=0.0), data_format='NCHW', name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())
x = paddle.to_tensor(data=np.random.randn(4, 50, 244, 244).astype('float32'))
l_o = paddle.nn.LayerNorm(normalized_shape=x.shape[1:], epsilon=1e-05,
    weight_attr=paddle.ParamAttr(learning_rate=0.0), bias_attr=paddle.
    ParamAttr(learning_rate=0.0), name=None)(x)
print(l_o.min(), l_o.max(), l_o.mean())

```

**Comment by justld** [comment]
image

File: paconvert/api_matcher.py [hunk 6]
```diff
@@ -508,3 +508,160 @@ def generate_code(self, kwargs):
...
+        if 'affine' in kwargs and 'False' in kwargs['affine']:
+            API_TEMPLACE = textwrap.dedent(
+                '''
+                paddle.nn.GroupNorm(num_groups={}, 
+                                    num_channels={}, 
+                                    epsilon={}, 
+                                    weight_attr=paddle.ParamAttr(learning_rate=0.0), 
+                                    bias_attr=paddle.ParamAttr(learning_rate=0.0), 
+                                    data_format='NCHW',
```
Comment (zhouwei25) [inline paconvert/api_matcher.py:550]: 这些name/data_format，不需要的参数就不用转写了。

可能存在个问题，如果用户没输入 `eps` 就会报错

**Review by zhouwei25 — APPROVED** [review]
